teps/0069-support-retries-for-custom-task-in-a-pipeline.md

```diff
@@ -0,0 +1,278 @@
+- [Requirements](#requirements)
+- [Proposal](#proposal)
+  - [Notes/Caveats (optional)](#notescaveats-optional)
+  - [Risks and Mitigations](#risks-and-mitigations)
+  - [User Experience (optional)](#user-experience-optional)
+  - [Performance (optional)](#performance-optional)
+- [Design Details](#design-details)
+- [Test Plan](#test-plan)
+- [Design Evaluation](#design-evaluation)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Infrastructure Needed (optional)](#infrastructure-needed-optional)
+- [Upgrade &amp; Migration Strategy (optional)](#upgrade--migration-strategy-optional)
+- [References (optional)](#references-optional)
+<!-- /toc -->
+
+## Summary
+
+A pipeline task can be configured with a `retries` count, this is 
+currently only supported for `TaskRun`s and not `Run`s (i.e. custom tasks).
+
+This TEP is about, a pipeline task can be configured with a `retries` count
+for Custom tasks.
+
+Also, a `PipelineRun` already manages a retry for regular task
+by updating it's status. However, for custom task, a tekton owned controller
+can signal a custom task controller, to retry. A custom task controller may
+optionally support it.
+
+## Motivation
+
+**Allow custom tasks to be configured with `task.retries`**
+
+Currently, a custom task controller has to develop its own retries support, which
+is not configurable as a pipeline task. It is true that not every custom task 
+need to support `retries`. For those who do want to support have to build
+their own solutions.
+
+There is no way to view retries information at the pipeline run level.
+
+In addition to building their own solutions, there is lack of uniformity in each
+custom task controller way of retries. This TEP will bring in standard/uniform way
+of supporting retry amongst custom controllers.
+
+As a side benefit, a custom task controller - developer SDK, might also benefit
+from this support, in the future, for example it can include documentation and
+stub code to make it easy how to support it.
+
+### Goals
+* Support propagating `pipelineSpec.task.retries` count information to custom-task
+  controllers.
+* Support signalling a `retry` to custom task controller, for a specific run. A
+  custom controller may optionally support it.
+* Gracefully handle the case where, custom controller does not support retry
+  and yet the `PipelineRun` happens to be configured with retry. This also
+  implies, an existing controller should not mis-behave if it is _not_ upgraded
+  to support retries.
+
+### Non-Goals
+* Directly, force update the `status.conditions` of a custom task.
+
+### Use Cases (optional)
+
+<!--
+Describe the concrete improvement specific groups of users will see if the
+Motivations in this doc result in a fix or feature.
+
+Consider both the user's role (are they a Task author? Catalog Task user?
+Cluster Admin? etc...) and experience (what workflows or actions are enhanced
+if this problem is solved?).
+-->
```

Comment:
Is it possible to add some use cases here describing why this is needed?

It almost makes sense to me without use cases but the part I am missing is that the way we do this with Pipeline -> TaskRun is that hte Pipeline controls the retries; I'd like to understand why we want the Pipeline to pass responsibility to the Run in this case, and why having a "retry" parameter for cases where it's important that the Custom Task do it's own retrying isn't enough

Reply:
Yes, I can add.

A pipelineRun will have the information of retry history of a Run, just as it has for a TaskRun. So Pipeline run is responsible for controlling the retry, e.g. the passing of `Retries` is FYI only. Actual, retry is triggered by pipelinerun controller, by setting `/spec/status` for a Run. However, how exactly the retry is done, is taken care of by the custom task controller. It may not respond at all, and we should be able to handle that.

Reply:
Thanks for the extra info - another alternative approach could be to have the Pipeline create a new Run every time a retry is needed, do you think there's any possibility that might work? I'm guessing that would have some downsides in that the custom task wouldn't have any control over how the retries work? 🤔 

(This reminds me of our other conversations in https://github.com/tektoncd/community/pull/422 in that I only recently realized we are re-using the same TaskRun when we retry; part of the motivation of decoupling TaskRuns from Pipelines was that something like a retry of a TaskRun could be accomplished by simply making as many TaskRuns as we wanted, but apparently thats not the route we went - so I'm hoping we can decide conclusively which approach we want to embrace)

Reply:
A custom task is different from a regular task, retrying can be very different from just retrying a Pod. So, a custom task may need to examine the state between two retries and optimise.

e.g. PipelineLoop controller can retry only failed pipelines based on certain criteria, instead of retrying everything incase we created a fresh Run.

Reply:
Can you explain a bit more about what kind of criteria the PipelineLoop controller might be looking at?

Reply:
A PipelineLoop controller maintains the list of failed pipeline runs for each Run. 
e.g. for a particular run - 2 out of 5 loops were not successful and as a result, it updates its status as Failed.

1. We give it a signal to retry, it can simply retry the two failed PipelineRun loops and not everything.

2. We create a fresh Run to retry, it has no information of what happened before. So, it starts as if fresh.

Reply:
Ahhh okay that makes a lot of sense - if you're willing to add this to the use cases description i think that would be great context to record.

Thanks!

Reply:
done, updated the alternatives section too!